pkg/gardenlet/controller/bastion/bastion_actuator.go

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	v1alpha1constants "github.com/gardener/gardener/pkg/apis/core/v1alpha1/constants"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/extensions"
+	extensionsbastion "github.com/gardener/gardener/pkg/operation/botanist/component/extensions/bastion"
+	"github.com/gardener/gardener/pkg/utils/flow"
+	gutil "github.com/gardener/gardener/pkg/utils/gardener"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/kubernetes/health"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	defaultTimeout         = 30 * time.Second
+	defaultSevereThreshold = 15 * time.Second
+	defaultInterval        = 5 * time.Second
+)
+
+// Actuator acts upon Bastion resources.
+type Actuator interface {
+	// Reconcile reconciles the Bastion.
+	Reconcile(context.Context) error
+	// Delete deletes the Bastion.
+	Delete(context.Context) error
+	// Migrate migrates the Bastion.
+	Migrate(context.Context) error
```

Comment:
I'm not too sure if should implement this for the `Bastion` resource... probably we shouldn't. I would suggest to enhance the `Bastion` controller in GCM (added via #4005) to auto-delete the `Bastion` resources referencing a shoot whose `.spec.seedName` changes. OK? 

cc @plkokanov @stoyanr

Reply:
Basically, if you do not want to keep the bastion instances alive during migration, you could do what @rfranzke suggested or delete the `operations.gardener.cloud.Bastion` resource during the shoot control plane migration flow - somewhere [here]( https://github.com/gardener/gardener/blob/master/pkg/gardenlet/controller/shoot/shoot_control_migrate.go#L153) so that the shoot namespace can be properly cleaned up.